# Практическая работа по дисциплине "Теория принятия решений"
## Тема: Применение методов линейного и динамического программирования для решения практических задач (по вариантам)
### Вариант 47

**Половников Никита**

**группа 1374**

## Задача 1
        В состав предприятия угольной промышленности входит 5 карьеров (шахт) и 3 пункта централизованного хранения угля, расположенных в одном регионе. Из пунктов централизованного хранения уголь поставляется на теплоэлектростанции, где используется в качестве топлива.
        Количество угля, доставляемое в пункт централизованного хранения из карьера, не должно превышать суточной добычи на данном карьере (табл. 1). Суммарное количество угля, поступающее со всех карьеров в пункты централизованного хранения, должно равняться суммарному объему ежедневных поставок потребителю (70 тонн). Каждый пункт централизованного хранения обладает максимальной емкостью (см. табл. 2).
        Угольная смесь, поставляемая потребителю, должна удовлетворять определенным ограничениям по составу: содержание золы, серы, влаги, летучих веществ (ограничения должны выполняться для каждого пункта централизованного хранения). Содержание компонентов в угле из каждого карьера, а также ограничения для каждого из компонентов в поставляемой потребителю смеси приведены в табл. 1. 
        Известна себестоимость добычи тонны угля в каждом из карьеров (табл. 1) и средние затраты на перевозку тонны угля от карьеров до пунктов централизованного хранения (табл. 3). 
        1. Определить план суточной добычи угля, обеспечивающий потребность в угле определенного качества и минимизирующий затраты предприятия. 
        2. Выявить пункты централизованного хранения, расширение которых способно снизить суточные затраты предприятия. Если такие пункты есть, то определить для них максимальную суточную стоимость дополнительной единицы емкости, при которой затраты будут меньше полученных в п. 1. 
        3. Выявить карьеры, увеличение суточной добычи на которых способно снизить суточные затраты. Если такие карьеры есть, определить максимальную возможную доплату за добычу «сверхплановой» тонны, позволяющую снизить затраты по сравнению с полученными в п. 1.

**Таблица 1: Характеристики карьеров**

| **Карьер** | **Суточная добыча, т** | **Себестоимость, ДЕ/т** | **Зольность, %** | **Сера, %** | **Влага, %** | **Летучие в-ва, %** |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| **1** | 16 | 13.6 | 22.6 | 2.5 | 8.9 | 31.3 |
| **2** | 23 | 12.4 | 14.3 | 1.8 | 8.7 | 36.1 |
| **3** | 25 | 18.5 | 18.3 | 1.2 | 7.4 | 30.7 | 
| **4** | 20 | 17.2 | 17.7 | 1.8 | 6.6 | 37.6 |  
| **5** | 19 | 20.4 | 21.5 | 2.1 | 7.4 | 30.0 | 

**Ограничения:** 

- Зольность $\leq 20$

- Сера $\leq 2$

- Влага $\leq 8$

- Летучие в-ва $\leq 35$ </br>

**Таблица 2: Ёмкость пунктов централизованного хранения**

| **Пункт** | **Ёмкость** |
|:---:|:---:|
| **1** | 12 |
| **2** | 24 |
| **3** | 54 |

</br>

**Таблица 3: Стоимость перевозок, ДЕ/т**

|  | **1** | **2** | **3** |
|:---:|:---:|:---:|:---:|
| **1** | 23.0 | 49.8 | 61.5 |
| **2** | 36.4 | 72.0 | 39.5 |
| **3** | 7.8 | 12.9 | 47.9 | 
| **4** | 36.6 | 18.6 | 39.8 |
| **5** | 38.3 | 62.5 | 36.6 |



### Формализация задачи 

*Решаем задачу минимизации затрат предприятия.*

**Целевая функция:**

#### $ \normalsize c = \sum{\sum{p_{ij}x_{ij}}} + \sum{r_{ij} \sum{x_{ij}}} \rightarrow min$,

где $\normalsize p_{ij}$ - стоимость перевозки из $i$-ой шахты в $j$-ый пункт хранения,

a $\normalsize r_{j}$ - себестоимость угля (ДЕ/т) $i$-ой шахты.


**Целевую функцию запишем в виде:**

$ \normalsize c = p_{11}x_{11} + p_{12}x_{12} + ... + p_{53}x_{53} + r_{1}(x_{11}+ x_{12}+x_{13} + ... + r_{5}(x_{51}+ x_{52}+x_{53}) \rightarrow min$

***Ограничения:***

$\bullet$ **По суточной добыче:**

$ x_{11}+ x_{12}+x_{13} \leq 16 $

...

$ x_{51}+ x_{52}+x_{53} \leq 19 $

$\bullet$ **По суммарному объему ежедневных поставок:**

$ x_{11}+ x_{12}+ ... + x_{53} = 70 $


$\bullet$ **По емкости пунктов:**

$ x_{11}+ x_{21} + ... +x_{51} \leq 12 $

$ x_{12}+ x_{22} + ... +x_{52} \leq 24 $

$ x_{13}+ x_{23} + ... +x_{53} \leq 54 $

$\bullet$ **По составу угля для каждого пункта хранения:**

- **По зольности:**

  **1 п.х:** $ 0.226 \cdot x_{11}+ 0.143 \cdot x_{21} + ... + 0.215 \cdot x_{51} \leq 12 \cdot 0.2$

  ...

  **3 п.х:**$ 0.226 \cdot x_{51}+ 0.143 \cdot x_{52}+ ...0.215 \cdot x_{53} \leq 54 \cdot 0.2 $ 
  
  
- **По сере:**

  **1 п.х:** $ 0.025 \cdot x_{11}+ 0.018 \cdot x_{21} + ... + 0.021 \cdot x_{51} \leq 12 \cdot 0.02$

  ...

  **3 п.х:**$ 0.025 \cdot x_{51}+ 0.018 \cdot x_{52}+ ... 0.021 \cdot x_{53} \leq 54 \cdot 0.02 $ 
  
- **По влаге:**

  **1 п.х:** $ 0.089 \cdot x_{11}+ 0.087 \cdot x_{21} + ... + 0.074 \cdot x_{51} \leq 12 \cdot 0.08$

  ...

  **3 п.х:**$ 0.089 \cdot x_{51}+ 0.087 \cdot x_{52}+ ... 0.074 \cdot x_{53} \leq 54 \cdot 0.08 $ 
  
- **По летучим в-вам:**

  **1 п.х:** $ 0.316 \cdot x_{11}+ 0.361 \cdot x_{21} + ... + 0.3 \cdot x_{51} \leq 12 \cdot 0.35$

  ...

  **3 п.х:**$ 0.316 \cdot x_{51}+ 0.361 \cdot x_{52}+ ... 0.3 \cdot x_{53} \leq 54 \cdot 0.35 $ 

### Описание задачи с помощью средств выбранной библиотеки

    1. Определить план суточной добычи угля, обеспечивающий потребность в угле определенного качества и минимизирующий затраты предприятия. 

In [6]:
# !pip install pulp
import warnings
warnings.filterwarnings("ignore")

In [1]:
from pulp import *

# Создаем задачу линейного программирования
prob = LpProblem("Minimize Cost", LpMinimize)

# Переменные решения
x = LpVariable.dicts("x", range(40), lowBound=0)

# Определяем целевую функцию
prob += 23.0*x[0] + 49.8*x[1] + 61.5*x[2] + 36.4*x[3] + 72.0*x[4] + 39.5*x[5] + 7.8*x[6] + 12.9*x[7] + 47.9*x[8] + 36.6*x[9] + 18.6*x[10] + 39.8*x[11] + 38.3*x[12] +  62.5*x[13] + 36.6*x[14] + 13.6*(x[0]+x[1]+x[2]) + 12.4*(x[3]+x[4]+x[5]) + 18.5*(x[6]+x[7]+x[8]) + 17.2*(x[9]+x[10]+x[11]) + 20.4*(x[12]+x[13]+x[14]), "Total Cost"

# Добавляем ограничения
prob += x[0]+ x[1] + x[2] <= 16, "Constraint 1" # усл. сут. добычи для 1 карьера
prob += x[3]+ x[4] + x[5] <= 23, "Constraint 2" # усл. сут. добычи для 2 карьера
prob += x[6]+ x[7] + x[8] <= 25, "Constraint 3" # усл. сут. добычи для 3 карьера
prob += x[9]+ x[10] + x[11] <= 20, "Constraint 4" # усл. сут. добычи для 4 карьера
prob += x[12]+ x[13] + x[14] <= 19, "Constraint 5" # усл. сут. добычи для 5 карьера

prob += x[0] + x[1] + x[2] + x[3] + x[4] + x[5] + x[6] + x[7] + x[8] + x[9] + x[10] + x[11] + x[12] + x[13] + x[14] == 70, "Constraint 6"

prob += 
x[0] + x[3] + x[6] + x[9] + x[12] <= 12, "Constraint 7" # усл. вметимости для 1 п.х.
prob += x[1] + x[4] + x[7] + x[10] + x[13] <= 24, "Constraint 8" # усл. вметимости для 2 п.х.
prob += x[2] + x[5] + x[8] + x[11] + x[14] <= 54, "Constraint 9" # усл. вметимости для 3 п.х.
prob += 0.226*x[0] + 0.143*x[3] + 0.183*x[6] + 0.177*x[9] + 0.215*x[12] <= 12*0.2, "Constraint 10" # усл. зольности для 1 п.х.
prob += 0.226*x[1] + 0.143*x[4] + 0.183*x[7] + 0.177*x[10] + 0.215*x[13] <= 24*0.2, "Constraint 11" # усл. зольности для 2 п.х.
prob += 0.226*x[2] + 0.143*x[5] + 0.183*x[8] + 0.177*x[11] + 0.215*x[14] <= 54*0.2, "Constraint 12" # усл. зольности для 3 п.х.

prob += 0.025*x[0] + 0.018*x[3] + 0.012*x[6] + 0.018*x[9] + 0.021*x[12] <= 12*0.02, "Constraint 13" # усл. зольности для 1 п.х.
prob += 0.025*x[1] + 0.018*x[4] + 0.012*x[7] + 0.018*x[10] + 0.021*x[13] <= 24*0.02, "Constraint 14" # усл. зольности для 2 п.х.
prob += 0.025*x[2] + 0.018*x[5] + 0.012*x[8] + 0.018*x[11] + 0.021*x[14] <= 54*0.02, "Constraint 15" # усл. зольности для 3 п.х.

prob += 0.089*x[0] + 0.087*x[3] + 0.074*x[6] + 0.066*x[9] + 0.074*x[12] <= 12*0.08, "Constraint 16" # усл. влаги для 1 п.х.
prob += 0.089*x[1] + 0.087*x[4] + 0.074*x[7] + 0.066*x[10] + 0.074*x[13] <= 24*0.08, "Constraint 17" # усл. влаги для 2 п.х.
prob += 0.089*x[2] + 0.087*x[5] + 0.074*x[8] + 0.066*x[11] + 0.074*x[14] <= 54*0.08, "Constraint 18" # усл. влаги для 3 п.х.

prob += 0.313*x[0] + 0.361*x[3] + 0.307*x[6] + 0.376*x[9] + 0.3*x[12] <= 12*0.35, "Constraint 19" # усл. летучих в-в для 1 п.х.
prob += 0.313*x[1] + 0.361*x[4] + 0.307*x[7] + 0.376*x[10] + 0.3*x[13] <= 24*0.35, "Constraint 20" # усл. летучих в-в для 2 п.х.
prob += 0.313*x[2] + 0.361*x[5] + 0.307*x[8] + 0.376*x[11] + 0.3*x[14] <= 54*0.35, "Constraint 21" # усл. летучих в-в для 3 п.х.


# Решаем задачу
prob.solve()

# Выводим результат
print("Status:", LpStatus[prob.status])
for v in prob.variables():
    print(v.name, "=", v.varValue)
print("Total Cost =", value(prob.objective))


Status: Optimal
x_0 = 0.0
x_1 = 0.0
x_10 = 11.0
x_11 = 9.0
x_12 = 0.0
x_13 = 0.0
x_14 = 2.0
x_2 = 0.0
x_3 = 0.0
x_4 = 0.0
x_5 = 23.0
x_6 = 12.0
x_7 = 13.0
x_8 = 0.0
x_9 = 0.0
Total Cost = 2938.3


D:\Users\nik_p\anaconda3\Lib\site-packages\pulp\pulp.py:1316: UserWarning: Spaces are not permitted in the name. Converted to '_'
  warnings.warn("Spaces are not permitted in the name. Converted to '_'")


**План суточной добычи угля, обеспечивающий потребность в угле определенного качества и минимизирующий затраты предприятия:**

- **2** карьер $\rightarrow$ **3** пункт хранения  $ \ $  **23** тонны

- **3** карьер $\rightarrow$ **1** пункт хранения $ \ $  **12** тонн

- **3** карьер $\rightarrow$ **2** пункт хранения $ \ $  **13** тонн

- **4** карьер $\rightarrow$ **2** пункт хранения $ \ $  **11** тонн

- **4** карьер $\rightarrow$ **3** пункт хранения $ \ $  **9** тонн

- **5** карьер $\rightarrow$ **3** пункт хранения $ \ $  **2** тонны

Стоимость = **2938.3** ДЕ

    2. Выявить пункты централизованного хранения, расширение которых способно снизить суточные затраты предприятия. Если такие пункты есть, то определить для них максимальную суточную стоимость дополнительной единицы емкости, при которой затраты будут  меньше полученных в п. 1.
    
    3. Выявить карьеры, увеличение суточной добычи на которых способно снизить суточные затраты. Если такие карьеры есть, определить максимальную возможную доплату за добычу «сверхплановой» тонны, позволяющую снизить затраты по сравнению с полученными в п. 1.

In [2]:
import pandas as pd
o = [{'name':name,'shadow price':c.pi,'slack': c.slack} for name, c in prob.constraints.items()]

pd.DataFrame(o)

name  shadow price   slack
0    Constraint_1           0.0  16.000
1    Constraint_2          -5.1  -0.000
2    Constraint_3          -4.4  -0.000
3    Constraint_4           0.0  -0.000
4    Constraint_5           0.0  17.000
5    Constraint_6          57.0  -0.000
6    Constraint_7         -26.3  -0.000
7    Constraint_8         -21.2  -0.000
8    Constraint_9           0.0  20.000
9   Constraint_10           0.0   0.204
10  Constraint_11           0.0   0.474
11  Constraint_12           0.0   5.488
12  Constraint_13           0.0   0.096
13  Constraint_14           0.0   0.126
14  Constraint_15           0.0   0.462
15  Constraint_16           0.0   0.072
16  Constraint_17           0.0   0.232
17  Constraint_18           0.0   1.577
18  Constraint_19           0.0   0.516
19  Constraint_20           0.0   0.273
20  Constraint_21           0.0   6.613

### В нашем случае shadow_prices 7 и 8  $!= 0$ $\rightarrow$ можно минимизировать затраты за счет ***расширения этих пунктов хранения.***
#### Для пункта централизованного хранения №1 максимальная суточная стоимость дополнительной единицы емкости должна быть меньше, чем 26.3 у. е.
#### Для пункта централизованного хранения №2 максимальная суточная стоимость дополнительной единицы емкости должна быть меньше, чем 21.2 у. е.
### В нашем случае shadow_prices 2 и 4  $!= 0$ $\rightarrow$ можно минимизировать затраты за счет ***увеличения суточной добычи на этих карьерах***
#### Для карьера №2 максимальная возможная доплата за добычу "сверхплановой" тонны должна быть меньше, чем 5.1 у. е.
#### Для карьера №3 максимальная возможная доплата за добычу "сверхплановой" тонны должна быть меньше, чем 4.4 у. е.
Constraint_1 (Ограничение 1): Значение теневой цены равно 0.0. Это означает, что увеличение или уменьшение этого ограничения не влияет на оптимальное значение целевой функции.

Constraint_2 (Ограничение 2): Значение теневой цены равно -5.1. Отрицательное значение означает, что увеличение добычи  на 1 единицу приведет к уменьшению значения целевой функции на 5.1.

и т.д.

## Вывод:

В ходе решения задали линейного програмиирования был найден оптимальный план добычи угля, минимизирующий затраты угля. Также проведен анализ чувствительности, по результатам которого, понятно какие пункты хранения можно расширить и на каких карьерах можно увеличить добычу для минимизации затрат.